Add WData plugin to database

Belyor opened the pull request:
### Description
- add test data data/WData_test_data.7z
- manual edit src/databases/CMakeLists.txt
- add database to /src/databases/WData
- add to gitignore .vscode files

### Type of change
* [ ] Bug fix
* [ ] New feature
* [ ] Documentation update
* [X] Other <!-- please explain with a note below -->
Adding WData plugin to visit database

### How Has This Been Tested?
I am using this plugin for over a year now in my daily work as a locally installed one on the server. It has been tested on Unix OS.

### Checklist:
<!-- For items in this checklist that do not apply, simply insert two tilde chars, `~~`, just ahead of the left bracket char, `[` at the beginning of a line. Each line ends with two tilde chars to make doing such ~~strikeouts~~ easy. -->

- [x] I have followed the [style guidelines][1] of this project. (Partly true. We are using the plugin as a library and include it to database as #include "wdata.c/.h" thus inside it we do NOT make any formatting changes)~~
- [x] I have performed a self-review of my own code.~~
- [x] I have commented my code where applicable.~~
- ~~[ ] I have updated the release notes.~~
- ~~[ ] I have made corresponding changes to the documentation.~~
- ~~[ ] I have added debugging support to my changes.~~
- [x] I have added tests that prove my fix is effective or that my feature works.~~
- ~~[ ] I have confirmed new and existing unit tests pass locally with my changes.~~
- ~~[ ] I have added new baselines for any new tests to the repo.~~
- [x] I have NOT made any changes to [*protocol* or *public interfaces*][3] in an RC branch.~~
- ~~[ ] I have assigned reviewers (see [VisIt's PR procedures][2] for more information).~~

[1]: https://visit-sphinx-github-user-manual.readthedocs.io/en/develop/dev_manual/StyleGuide.html
[2]: https://visit-sphinx-github-user-manual.readthedocs.io/en/develop/dev_manual/pr_create.html#reviewers
[3]: https://visit-sphinx-github-user-manual.readthedocs.io/en/develop/dev_manual/RCDevelopment.html#communication-protocols-and-public-apis

CYB3R42Y committed:
WData plugin init commit

- test data added _data/WData_test_data.7z_
- manual edit _src/databases/CMakeLists.txt_
- add database to _/src/databases/WData_
- add to gitignore _.vscode_ file

biagas reviewed (changes requested):
Copyright notice needs updating.

biagas commented on src/databases/CMakeLists.txt:
```diff
@@ -379,6 +382,7 @@ if(VISIT_PARADIS)
 endif()
 
 CHECK_THIRDPARTY_SATISFIED(H5Part H5PART)
+CHECK_THIRDPARTY_SATISFIED(WData WData)
```
Our CHECK_THIRDPARTY_SATISFIED macro is for checking if third-party (external) dependencies have been met. It looks to me like you included WData source with the plugin and that there are no external dependencies. Have I read that correctly?  If there are no external dependencies, this line can be removed and your plugin added to the REQUIRED_DATABASE_PLUGINS list lower down in this file.

Belyor replied:
This is something I was not sure, as some of the databases uses third-party names as their names. This was edited manually. To be clear, we do not use any other dependencies

Belyor replied:
I have included the CMakeLists.txt as it is. File is generated on our local server. After the commit pipeline tests failed. I suppose that its due to this part:

> CMAKE_MINIMUM_REQUIRED(VERSION 3.8 FATAL_ERROR)
> SET(VISIT_PLUGIN_DIR "/home/makowski/.visit/3.1.2/linux-x86_64/plugins")
> INCLUDE("/usr/local/visit-3.1.2/3.1.2/linux-x86_64/include/PluginVsInstall.cmake")
> INCLUDE("/usr/local/visit-3.1.2/3.1.2/linux-x86_64/include/VisItLibraryDependencies.cmake")

How should I follow?

biagas commented on src/databases/WData/wdata.xml:
```diff
@@ -0,0 +1,14 @@
+<?xml version="1.0"?>
+  <Plugin name="WData" type="database" label="WDATA format reader" version="0.2.0" enabled="true" mdspecificcode="false" engspecificcode="false" onlyengine="false" noengine="false" dbtype="MTSD" haswriter="false" hasoptions="false" filePatternsStrict="false" opensWholeDirectory="false">
+    <CXXFLAGS>
+      -I./
```
Adding the current directory as an include isn't required. It is handled automatically when the CMakeLists.txt is generated by 'xml2cmake'.

Belyor replied:
Thats easy enough to fix

Belyor replied:
Since I can not comment on this [response](https://github.com/visit-dav/visit/pull/17033#pullrequestreview-758766619) I will comment here.
We decided to edit _Visit_ namings to **WData**, but the library functions (see all functions in wdata.h) will still be formatted with lowercase. It is due that we have a set of tools that uses this library and changing it would produce a lot of errors. From now on all classes, methods, and others that are used in or by Visit, we name **WData**, otherwise all **wdata** lib functions are with lowercase.

biagas replied:
That is perfectly reasonable, thanks for clarifying. It had not been my intention to suggest changing the library function names.

biagas reviewed (changes requested):
I noticed an inconsistency in use of 'wdata' vs 'WData'. Both in #includes and in filenames.
I suggest using 'WData' since that is what you named your directory and most of the files.
You can run 'xml2info -clobber WData.xml' to regenerate the Info files. This should fix the copyright notice as well as the includes for the Info files.
Also, please run 'xml2cmake' and add the CMakeLists.txt file to the repo. We require them to exist in the repo, but future maintenance is performed automatically. I apologize that I didn't make that part clear in our discussion.

brugger1 commented on src/databases/WData/avtwdataFileFormat.C:
```diff
@@ -0,0 +1,861 @@
+#include "wdata.c"
+
+#define wdata_operation(cmd)                                                       \
+    {                                                                              \
```
Style nitpick: This block of code doesn't need to be indented.

Belyor replied:
This is something that C formatter suggests. I can manually get rid of it if that is something You insist to do

brugger1 replied:
I didn't know that. Then leaving it as is sounds like the right thing to do.

brugger1 commented on src/databases/WData/avtwdataFileFormat.C:
```diff
@@ -0,0 +1,861 @@
+}
+
+// =======================================================================================
+// ============================= WDataComplexVariable =======================================
```
Style nitpick: There are 3 extra "=" at the end of this line.

Belyor replied:
Done

brugger1 commented on src/databases/WData/avtwdataFileFormat.C:
```diff
@@ -0,0 +1,861 @@
+}
+
+// =======================================================================================
+// ============================= WDataVectorVariable =======================================
```
Style nitpick: There are 2 extra "=" at the end of the line.

Belyor replied:
Done

CYB3R42Y committed:
PR review part 1

- WData copyright update
- update database _CMakeLists.txt_
- small formatting tweaks
- add _CMakeLists.txt_ to WData database path
- edit WData.xml file
- consistency in WData/wdata namings

markcmiller86 commented on src/databases/WData/WDataCommonPluginInfo.C:
```diff
@@ -0,0 +1,109 @@
+/*****************************************************************************
```
I am not sure where lines 1...36 came from here and in a few other source files in this PR? Was it cut-n-paste from another source file? Or, did our `xml2plugin` tools wind up producing that?

In any case, please cut that down to the same 3-line copyright used elsewhere as in for example (totally random selection of a database plugin source file), https://github.com/visit-dav/visit/blob/f060e01861a317ce327623b85039251933b3d26b/src/databases/Image/avtImageFileFormat.C#L1-L3

I think it is perfectly fine to also add your own copyright statement if you would like, after ours here too and I would encourage you to do that if you are so inclined.

Belyor replied:
The lines (1-36) are what **xml2** tool produced. I might use the old version by mistake (we have a visit 2.13 module on the local server) and that is the reason for that. Also, I used **xml2cmake** if that makes any difference. 
Included our licence to wdata.c/.h files

markcmiller86 commented:
I noticed spacing and position of some of the markdown in the opening comment was defeating proper rendering. I've corrected that. For future ref, to mark an item as not-applicable, you add the squiggles `~~` *after* the list item indicator, `-`. And to mark an item checked, use lower-case `x` with no spaces betwee `[` and `]` (or alternatively, just check it with the mouse once the issue or PR has been submitted.

CYB3R42Y committed:
PR review part 2

- Copyright update: add GNU to wdata.c/.h files and shorten the visit files autogenerated

CYB3R42Y committed:
File naming changed

- consistency with WData/wdata file namings

CYB3R42Y committed:
Edit plugin CMakeLists.txt

- manually edit CmakeLists.txt plugin file to be similar as the one in other databases

CYB3R42Y committed:
Update plugin CMakeLists.txt

Autogenerated CMakeLists.txt file in plugin path have had an additional folder in the include path

Belyor commented:
I think the code is ready for further review. Today I was working on implementing correct _CMakeLists.txt_ file in plugin path and believe that it works.
What is next step in the process?

biagas commented:
> What is next step in the process

Next step is to wait for the reviewers to have time to review the latest changes.  Once there are approvals, I can perform the merge for you.

Thank you again for contributing your plugin to VisIt.

biagas commented on src/databases/WData/wdata.c:
```diff
@@ -0,0 +1,1056 @@
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+#include <unistd.h>
```
There will be a problem compiling on Windows due to use of unistd and libgen headers.
I should be able to make the necessary changes after the merge of this PR if you don't have access to Windows machines yourself. (We use Visual Studio to compile VisIt on Windows).

Belyor replied:
I did not want to be rude, it's my first PR and this experience is totally new for me.
No one in our group has access to a Windows machine and Your help would be very helpful

biagas replied:
It is not rude at all, I understand not everyone has access to Windows.

brugger1 reviewed (approved):
Looks good!

biagas commented:
@markcmiller86 I cannot see how to resolve your requested changes. Can you take a look. I believe this is ready to merge.

markcmiller86 commented:
> @markcmiller86 I cannot see how to resolve your requested changes. Can you take a look. I believe this is ready to merge.

Hmmm...not sure what you might be talking about here beyond my request (which may have repeated yours) to update the copyright notices to the 3-line version, which I see now has been completed. The other [comment](https://github.com/visit-dav/visit/pull/17033#issuecomment-923578844) I left regarding spacing and position in markdown was *not* about the code here. It was about the formatting in the [initial comment](https://github.com/visit-dav/visit/pull/17033#issue-1000794628) used to open this PR. It was improperly formatted and I fixed that myself.
